### DISCUSSION TITLE
3.13.: All REST requests seem to yield "Unhandled exception in router" without any further info

### DISCUSSION THREAD

**Discussion by famod** [quarkusio/quarkus#42636] [Q&A]
In certain late stages of my project I'm getting:
> ERROR [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-4) Unhandled exception in router

without any further info/trace. This seems to happen for any REST request aimed towards this Quarkus service.
Scheduled jobs are fine though and I also got past my initial Liquibase issue (see https://issues.apache.org/jira/browse/LANG-1748), so the service is not broken entirely.

Any clue what's going on or any ideas how to get more details? I had a look at the vert.x class involved and I didn't see any debug or trace statements of much value, but I might be wrong.

**Comment by famod** [comment]
@cescoffier maybe? Thanks!

**Reply by cescoffier** [reply]
That means that a route throws an exception. Do you have further details?

**Reply by famod** [reply]
> Do you have further details?

Well, not really. Nothing else in the logs.
I merely updated from 3.12.2 to 3.13.2. The old release is still working fine on the affected systems/infrastructure.
The Quarkus service is receiving REST requests via HTTP from a HashiCorp Connect sidecard that is in turned called by haproxy via HTTPS.
Not only our regular REST APIs are affected, the same goes for health checks.

**Reply by gsmet** [reply]
For the only occurrence of this message I could find, there should  be a stacktrace... It looks like the `failure` throwable that should be there is `null` - which will probably makes things very hard to debug :/

**Reply by gsmet** [reply]
@famod if you can reproduce this easily, it would be a good idea to try to narrow down the version that introduced the regression.

@tsegismont any idea how we could get more information?

**Reply by cescoffier** [reply]
@famod any idea how we could reproduce it?

At least try with ` quarkus.log.category."io.vertx".min-level=TRACE`

**Reply by gsmet** [reply]
One thing you might also try would be to downgrade Vert.x to 4.5.7 (the version used in 3.12) and see how it goes.

**Reply by famod** [reply]
So I basically went back to before #41155 (but via depMgmt in my project, not via rebuilding Quarkus itself) and that fixed it!
That also includes a downgrade of netty - not sure how important that is.

I'll gather some logs next.

**Reply by famod** [reply]
@cescoffier 
> @famod any idea how we could reproduce it?

No clue for now. On our very early dev stages which don't use any HashiCorp stuff (especially no Consul Connect), everything is fine!

**Reply by cescoffier** [reply]
Hum, that's interesting. I do not see a good reason for failing like this. 

@Vietj, any idea?

**Reply by tsegismont** [reply]
I can't see anything in the [changes for Vert.x Web for 4.5.8](https://github.com/vert-x3/vertx-web/milestone/76?closed=1) that may lead to this.

If you can't provide a reproducer, the only way forward is to run your application in debug mode and inspect what happens in `io.vertx.ext.web.impl.RoutingContextImplBase#unhandledFailure`. Then please report your findings.

**Reply by gsmet** [reply]
I wouldn't be surprised if it was some low level Netty stuff. @cescoffier is it possible to only adjust the Vert.x version (either by running the old Vert.x with the new Netty or the opposite)?

**Reply by cescoffier** [reply]
Yes, for these versions, it should be possible adjust Netty and Vertx versions independently.

**Reply by famod** [reply]
> At least try with ` quarkus.log.category."io.vertx".min-level=TRACE`

Snip:
```
15:55:52,745 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.imp.RouterImpl] (vert.x-eventloop-thread-2) Router: 859016606 accepting request GET http://172.x.y.z:2080/q/health/live
15:55:52,751 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Route matches: RouteState{metadata=null, path='null', name=null, order=-200, enabled=true, methods=null, consumes=null, emptyBodyPermittedWithConsumes=false, produces=null, contextHandlers=[io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$AuthenticationHandler@34394a1d], failureHandlers=null, added=true, pattern=null, groups=null, useNormalizedPath=true, namedGroupsInRegex=null, virtualHostPattern=null, pathEndsWithSlash=false, exclusive=false, exactPath=true}
15:55:52,752 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Calling the  handler
15:55:52,791 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Route matches: RouteState{metadata=null, path='null', name=null, order=-199, enabled=true, methods=null, consumes=null, emptyBodyPermittedWithConsumes=false, produces=null, contextHandlers=[io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder$13@6a5fdd0e], failureHandlers=null, added=true, pattern=null, groups=null, useNormalizedPath=true, namedGroupsInRegex=null, virtualHostPattern=null, pathEndsWithSlash=false, exclusive=false, exactPath=true}
15:55:52,791 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Calling the  handler
15:55:52,792 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Route matches: RouteState{metadata=null, path='null', name=null, order=-100, enabled=true, methods=null, consumes=null, emptyBodyPermittedWithConsumes=false, produces=null, contextHandlers=[io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$1@5b207444], failureHandlers=null, added=true, pattern=null, groups=null, useNormalizedPath=true, namedGroupsInRegex=null, virtualHostPattern=null, pathEndsWithSlash=false, exclusive=false, exactPath=true}
15:55:52,792 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Calling the  handler
15:55:52,793 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Route matches: RouteState{metadata=null, path='null', name=null, order=-2, enabled=true, methods=null, consumes=null, emptyBodyPermittedWithConsumes=false, produces=null, contextHandlers=[io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$1@3af56774], failureHandlers=null, added=true, pattern=null, groups=null, useNormalizedPath=true, namedGroupsInRegex=null, virtualHostPattern=null, pathEndsWithSlash=false, exclusive=false, exactPath=true}
15:55:52,794 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Calling the  handler
15:55:52,794 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Route matches: RouteState{metadata=null, path='/q/', name=null, order=3, enabled=true, methods=null, consumes=null, emptyBodyPermittedWithConsumes=false, produces=null, contextHandlers=[io.vertx.ext.web.impl.RouteImpl$$Lambda/0x0000763a2cb78000@386df085], failureHandlers=[io.vertx.ext.web.impl.RouteImpl$$Lambda/0x0000763a2cb78220@46248a5], added=true, pattern=null, groups=null, useNormalizedPath=true, namedGroupsInRegex=null, virtualHostPattern=null, pathEndsWithSlash=true, exclusive=true, exactPath=false}
15:55:52,795 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Calling the  handler
15:55:52,796 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Route matches: RouteState{metadata=null, path='/health/live', name=null, order=4, enabled=true, methods=null, consumes=null, emptyBodyPermittedWithConsumes=false, produces=null, contextHandlers=[io.quarkus.smallrye.health.runtime.SmallRyeLivenessHandler@6e459102], failureHandlers=null, added=true, pattern=null, groups=null, useNormalizedPath=true, namedGroupsInRegex=null, virtualHostPattern=null, pathEndsWithSlash=false, exclusive=false, exactPath=true}
15:55:52,796 TRACE (247617c4448ea1e3a7011f38bb399ffd,a461ad835e013546) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-2) Calling the  handler
15:55:53,519 TRACE (91d4bfbda7a9a300155542fb417110da,f13e8a3aeea610f1) [io.ver.ext.web.imp.RouterImpl] (vert.x-eventloop-thread-0) Router: 859016606 accepting request GET http://0.0.0.0:2080/q/health/live
15:55:53,519 TRACE (91d4bfbda7a9a300155542fb417110da,f13e8a3aeea610f1) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-0) Route matches: RouteState{metadata=null, path='/q/', name=null, order=3, enabled=true, methods=null, consumes=null, emptyBodyPermittedWithConsumes=false, produces=null, contextHandlers=[io.vertx.ext.web.impl.RouteImpl$$Lambda/0x0000763a2cb78000@386df085], failureHandlers=[io.vertx.ext.web.impl.RouteImpl$$Lambda/0x0000763a2cb78220@46248a5], added=true, pattern=null, groups=null, useNormalizedPath=true, namedGroupsInRegex=null, virtualHostPattern=null, pathEndsWithSlash=true, exclusive=true, exactPath=false}
15:55:53,519 TRACE (91d4bfbda7a9a300155542fb417110da,f13e8a3aeea610f1) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-0) Calling the failure handler
15:55:53,520 TRACE (91d4bfbda7a9a300155542fb417110da,f13e8a3aeea610f1) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-0) Route matches: RouteState{metadata=null, path='null', name=null, order=2147483646, enabled=true, methods=null, consumes=null, emptyBodyPermittedWithConsumes=false, produces=null, contextHandlers=null, failureHandlers=[io.quarkus.resteasy.reactive.server.runtime.observability.ObservabilityIntegrationRecorder$1@53225219], added=true, pattern=null, groups=null, useNormalizedPath=true, namedGroupsInRegex=null, virtualHostPattern=null, pathEndsWithSlash=false, exclusive=false, exactPath=true}
15:55:53,520 TRACE (91d4bfbda7a9a300155542fb417110da,f13e8a3aeea610f1) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-0) Calling the failure handler
15:55:53,520 TRACE (91d4bfbda7a9a300155542fb417110da,f13e8a3aeea610f1) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-0) Route matches: RouteState{metadata=null, path='null', name=null, order=2147483647, enabled=true, methods=null, consumes=null, emptyBodyPermittedWithConsumes=false, produces=null, contextHandlers=null, failureHandlers=[io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder$7@323155c4], added=true, pattern=null, groups=null, useNormalizedPath=true, namedGroupsInRegex=null, virtualHostPattern=null, pathEndsWithSlash=false, exclusive=false, exactPath=true}
15:55:53,520 TRACE (91d4bfbda7a9a300155542fb417110da,f13e8a3aeea610f1) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-0) Calling the failure handler
15:55:53,520 TRACE (91d4bfbda7a9a300155542fb417110da,f13e8a3aeea610f1) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-0) Route matches: RouteState{metadata=null, path='null', name=null, order=2147483647, enabled=true, methods=null, consumes=null, emptyBodyPermittedWithConsumes=false, produces=null, contextHandlers=null, failureHandlers=[io.quarkus.vertx.http.runtime.QuarkusErrorHandler@2a5aa9c], added=true, pattern=null, groups=null, useNormalizedPath=true, namedGroupsInRegex=null, virtualHostPattern=null, pathEndsWithSlash=false, exclusive=false, exactPath=true}
15:55:53,520 TRACE (91d4bfbda7a9a300155542fb417110da,f13e8a3aeea610f1) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-0) Calling the failure handler
15:55:53,521 ERROR (,) [io.ver.ext.web.RoutingContext] (vert.x-eventloop-thread-0) Unhandled exception in router
```

Next I'll play around with just the netty version.

**Reply by famod** [reply]
It _is_ vertx 4.5.8!
- dep managed everything related to the state _before_ #41155 :+1: 
- as above, but only `vertx-dependencies` BOM from 4.5.7 to 4.5.8 :-1: 

Consequently, downgrading only netty (and quarkus-http) and leaving mutiny on > 4.5.7 had no effect.

**Reply by cescoffier** [reply]
Thanks. At least we can reduce the scope.

@tsegismont can you check with @Vietj?

**Reply by cescoffier** [reply]
According to the log, it could be something in ResteasyReactiveRecorder

**Reply by gsmet** [reply]
I looked at the commits from 4.5.8 and the only thing that could maybe cause issues could be: https://github.com/vert-x3/vertx-web/commit/50e054345b347e8e99d9f5024522801ea1e80236#diff-dbf88e0bf79f27956c1699c4894506bc87342e36578440f5c0ea2b20c24ad6f1R194-R197

But that's a wild guess.

**Reply by famod** [reply]
I was about to bisect this issue in vert.x core but man that's quite a zoo of modules - nothing for a late tuesday evening.
Tomorrow I will at least try to call the Quarkus service without anything in between (my ops guy for that was too busy today), to narrow it down further.

**Reply by cescoffier** [reply]
@gsmet that would produce 400 response, not the empty exception we are seeing. However that new code may throw an npe or something. 

@famod it would be great to have an idea of the request sent to the application (especially the headers)

**Reply by gsmet** [reply]
@famod given it’s a late micro, maybe the changes are minimal enough that you can try adjusting the version module per module so that you can get the problematic one. Then the bisect would be easier.

**Reply by famod** [reply]
@cescoffier 
> @famod it would be great to have an idea of the request sent to the application (especially the headers)

Yeah, will try!

@gsmet I think the main issue are the modules depending on core when you bisect in core. Anyway, I might be able to take another stab at bisecting tomorrow evening.

**Reply by tsegismont** [reply]
> @gsmet that would produce 400 response, not the empty exception we are seeing. However that new code may throw an npe or something.

@cescoffier yes, the purpose of this commit is to send a 400 response when the host header is missing. But @gsmet may be right, in the sense that it introduces a case where a NPE is thrown.

I see a difference in the logs sent by @famod : the first request that goes well is sent to `http://172.x.y.z:2080/q/health/live`, while the second that goes wrong is sent to `http://0.0.0.0:2080/q/health/live`.

I don't know why these requests are different (perhaps @famod can elaborate) but it would be worth trying if the problem only happens when the host is `0.0.0.0`.

**Reply by famod** [reply]
So I have (somewhat good) news: I realized that haproxy is responding with 503 instead of forwarding any request to the Quarkus service (via Consul Connect sidecar), because the health check requests it issues itself are running into this issue. This 503 behaviour works as designed.
Now comes the interesting part: For testing purposes, I removed the health checks from the haproxy config so that it doesn't block actual requests with 503 and voilà, no issues anymore! This also includes health check requests from other sources than haproxy itself, e.g. our monitoring system or manual request from e.g. my local browser. It's especially interesting that these request go all the way through haproxy -> Consul Connect -> Quarkus service, *so the issue seems to be specifically how the haproxy health check requests are looking like!*

This is still a showstopper for my setup because we can't just get rid of the health checks in/from haproxy, but it might explain why no other user has reported this issue (AFAICS).

I will try to capture a health check request issued by haproxy, but my ops guy is still busy elsewhere - so this will take more time.

**Reply by famod** [reply]
I was able to capture the request headers on my own via Quarkus access logging (`long` format) and **I think this prooves the host header thesis!**

All **ok**:
```
GET /q/health/live HTTP/1.1
Host: 172.x.y.z:2080
User-Agent: Consul Health Check
Accept: text/plain, text/*, */*
Accept-Encoding: gzip
Connection: close
```
```
GET /q/health/ready HTTP/1.1
user-agent: check_http/v2.3.1 (monitoring-plugins 2.3.1)
host: <redacted>
x-forwarded-proto: https
x-forwarded-for: <redacted>
```

```
GET /q/health/ready HTTP/1.1
host: <redacted>
user-agent: curl/7.61.1
accept: */*
x-forwarded-proto: https
x-forwarded-for: <redacted>
```

```	
GET /q/health/ HTTP/1.1
host: <redacted>
user-agent: Mozilla/5.0 (X11; Linux x86_64; rv:129.0) Gecko/20100101 Firefox/129.0
accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/png,image/svg+xml,*/*;q=0.8
accept-language: en-US,en;q=0.5
accept-encoding: gzip, deflate, br, zstd
upgrade-insecure-requests: 1
sec-fetch-dest: document
sec-fetch-mode: navigate
sec-fetch-site: cross-site
sec-fetch-user: ?1
priority: u=0, i
te: trailers
x-forwarded-proto: https
x-forwarded-for: <redacted>
```
(me via my browser)

And _drumroll_ **failing**:
```
GET /q/health/live HTTP/1.0
-
```

**Reply by famod** [reply]
So I was able to fix this issue by configuring the haproxy http-check to include the host header.
But then again, isn't the host header only mandator from http version 1.1 upwards but optional in 1.0?

**Reply by gsmet** [reply]
Yeah, my guess is that the changes in this patch were a bit too optimistic.

Now I think we need to fix two things:
- understand why we don't get a better error - because not having a clue of what's going on is not a good idea
- fix the issue if we think supporting HTTP 1.0 is a good idea - I'm pretty sure some low level stuff is still doing HTTP 1.0 requests

**Reply by cescoffier** [reply]
First, we want to support HTTP/1.0.

The bug is that the forwarded considered "host" mandatory for both HTTP/1.1 (expected) and HTTP/1.0 (bug) requests.

**Reply by cescoffier** [reply]
@tsegismont Can you see with @vietj to fix Vert.x Web?

**Reply by famod** [reply]
> understand why we don't get a better error - because not having a clue of what's going on is not a good idea

\+ :100: 

In that context, I noticed that:
a) access logging doesn't work if this issue strikes (so I had to capture the headers with Quarkus 3.12.2)
b) access logging doesn't have trace id / span id
c) the "Unhandled exception in router" error is also logged without trace id / span id

Maybe none of these issues are solvable because it all happens on such a low level...?

**Reply by cescoffier** [reply]
I guess it's an NPE - and in this case, we should have a more specific log message (with the stack trace).

**Reply by tsegismont** [reply]
> @tsegismont Can you see with @vietj to fix Vert.x Web?

Sure

**Reply by tsegismont** [reply]
For the record, I filed https://github.com/vert-x3/vertx-web/issues/2638 and Julien will look into it asap.
